File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        int n=s.size();
        string f=s;
        reverse(f.begin(),f.end());
        vector<vector<int>> dp( n+1 , vector<int> (n+1, 0));
        for(int i=1;i<n+1;++i)
            for(int j=1;j<n+1;++j){
                if(s[i-1]==f[j-1])
                    dp[i][j]=dp[i-1][j-1]+1;
                else
                    dp[i][j]=max(dp[i-1][j],max(dp[i][j-1],dp[i-1][j-1]));
            }
        return dp[n][n];
    }
};

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n=len(s)
        h=s[::-1]
        dp=[[0 for i in range(n+1)] for j in range(n+1)]
        for i in range(1,n+1):
            for j in range(1,n+1):
                if s[i-1]==h[j-1]:
                    dp[i][j]=dp[i-1][j-1]+1
                else:
                    dp[i][j]=max(dp[i-1][j-1],dp[i-1][j],dp[i][j-1])
       
        return dp[-1][-1]